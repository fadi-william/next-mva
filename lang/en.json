{
  "nav": {
    "home": "Home",
    "about": "About"
  },
  "lang": {
    "french": "French",
    "english": "English"
  }
}