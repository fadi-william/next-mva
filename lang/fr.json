{
  "nav": {
    "home": "Accueil",
    "about": "À props"
  },
  "lang": {
    "french": "Français",
    "english": "Anglais"
  }
}