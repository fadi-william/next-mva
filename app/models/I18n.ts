// This is a mst store to manage i18n in the application.
import { applySnapshot, types } from "mobx-state-tree";
import { addLocaleData } from "react-intl";

// Import the i18n local storage module.
import { getCurrentUserLocale, setCurrentUserLocale } from "../storage/i18n";

// Currently, we support english and french.
import * as enLocaleData from "react-intl/locale-data/en";
import * as frLocaleData from "react-intl/locale-data/fr";

// Import the i18n json files.
import * as enTranslatedData from "../../lang/en.json";
import * as frTranslatedData from "../../lang/fr.json";

export type TI18n = typeof I18n.Type;

// The module's global variables.
let store: TI18n = null;
let currentLocale = null;

export const I18n = types
  .model({
    locale: types.string,
    messages: types.optional(types.frozen, {})
  })
  .actions(self => ({
    loadLocale(lang: string) {
      self.locale = lang;
      setCurrentUserLocale(lang);
      if (lang === "en") {
        addLocaleData(enLocaleData);
        self.messages = enTranslatedData;
      } else {
        addLocaleData(frLocaleData);
        self.messages = frTranslatedData;
      }
    }
  }))
  .actions(self => ({
    afterCreate() {
      self.loadLocale(self.locale);
    }
  }));

// Utility function - Instantiates a part of the mobx-state-tree store.
export function getStore(locale?, snapshot = null) {
  if (!locale) {
    locale = getCurrentUserLocale();
    currentLocale = locale;
  }
  if (store === null || currentLocale !== locale) {
    store = I18n.create({ locale });
  }
  if (snapshot) {
    applySnapshot(store, snapshot);
  }
  return store;
}
