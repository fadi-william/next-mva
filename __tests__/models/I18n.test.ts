import { getSnapshot } from "mobx-state-tree";
import { getStore, I18n } from "../../app/models/I18n";

// Import the i18n json files.
import * as enTranslatedData from "../../lang/en.json";
import * as frTranslatedData from "../../lang/fr.json";

// Mock the i18n local storage module.
jest.mock("../../app/storage/i18n");

const i18nObj = {
    locale: "en",
    messages: enTranslatedData,
};

describe("i18n model tests", () => {

    it("can create an i18n model", () => {
        const i18nStore = I18n.create(i18nObj);

        expect(i18nStore.messages).toEqual(enTranslatedData);
        expect(getSnapshot(i18nStore)).toMatchSnapshot();
    });

    it("can change language", () => {
        const i18nStore = I18n.create(i18nObj);
        i18nStore.loadLocale("fr");

        expect(i18nStore.messages).toEqual(frTranslatedData);
        expect(getSnapshot(i18nStore)).toMatchSnapshot();
    });

    it("can get a store", () => {
        const locale = "en";
        const i18nStore = getStore(locale);

        expect(i18nStore.messages).toEqual(enTranslatedData);
        expect(getSnapshot(i18nStore)).toMatchSnapshot();
    });

    it("can restore am i18n model from a snaphost", () => {
        const locale = "en";
        const i18nStore = getStore(locale, i18nObj);

        expect(i18nStore.messages).toEqual(enTranslatedData);
        expect(getSnapshot(i18nStore)).toMatchSnapshot();
    });
});
