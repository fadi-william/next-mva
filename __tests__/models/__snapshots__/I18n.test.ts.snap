// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`i18n model tests can change language 1`] = `
Object {
  "locale": "fr",
  "messages": Object {
    "default": Object {
      "lang": Object {
        "english": "Anglais",
        "french": "Français",
      },
      "nav": Object {
        "about": "À props",
        "home": "Accueil",
      },
    },
    "lang": Object {
      "english": "Anglais",
      "french": "Français",
    },
    "nav": Object {
      "about": "À props",
      "home": "Accueil",
    },
  },
}
`;

exports[`i18n model tests can create an i18n model 1`] = `
Object {
  "locale": "en",
  "messages": Object {
    "default": Object {
      "lang": Object {
        "english": "English",
        "french": "French",
      },
      "nav": Object {
        "about": "About",
        "home": "Home",
      },
    },
    "lang": Object {
      "english": "English",
      "french": "French",
    },
    "nav": Object {
      "about": "About",
      "home": "Home",
    },
  },
}
`;

exports[`i18n model tests can get a store 1`] = `
Object {
  "locale": "en",
  "messages": Object {
    "default": Object {
      "lang": Object {
        "english": "English",
        "french": "French",
      },
      "nav": Object {
        "about": "About",
        "home": "Home",
      },
    },
    "lang": Object {
      "english": "English",
      "french": "French",
    },
    "nav": Object {
      "about": "About",
      "home": "Home",
    },
  },
}
`;

exports[`i18n model tests can restore am i18n model from a snaphost 1`] = `
Object {
  "locale": "en",
  "messages": Object {
    "default": Object {
      "lang": Object {
        "english": "English",
        "french": "French",
      },
      "nav": Object {
        "about": "About",
        "home": "Home",
      },
    },
    "lang": Object {
      "english": "English",
      "french": "French",
    },
    "nav": Object {
      "about": "About",
      "home": "Home",
    },
  },
}
`;
